lib/Sema/TypeCheckProtocolInference.cpp

```diff
@@ -444,6 +445,20 @@ AssociatedTypeInference::inferTypeWitnessesViaValueWitnesses(
     if (auto func = dyn_cast<FuncDecl>(req)) {
       if (func->isOperator() || isa<AccessorDecl>(func))
         continue;
+      auto containsDependentType = [&] {
+        for (const auto *param : *func->getParameters()) {
+          const auto paramType = param->getType();
```

Comment:
I suggest using getInterfaceType() instead of getType(). Instead of looking for an ArchetypeType, see if you have a DependentMemberType. Then check if the DependentMemberType's associated type, obtained by calling getAssocType(), is equal to `assoc`. This will replace the identifier string comparison with something more rigorous.

Also, you should do a recursive check if the type contains any dependent member types, since they might appear but not at the top level of the type, for example, `[X]` is an array type containing `X`. If you add a test case involving a structural type you will see that it will fail with the current implementation.

To search a type recursively, use `type.findIf()`.

Reply:
I've found some issue of my implementation that can't infer associcatedtype from method signature. So my approach is incorrect and I updated PR.

```swift
protocol P {
    associatedtype X = Int

    func f1(_ x: X)
    func f2(_ x: X)
}

struct S: P {
    func f1(_ x: Int) {} // error: reference to invalid associated type 'X' of type 'S'
    func f2(_ x: X) {}
}